crates/astria-sequencer/src/grpc/optimistic.rs

```diff
@@ -0,0 +1,152 @@
+        GetBlockCommitmentStreamResponse,
+        GetOptimisticBlockStreamRequest,
+        GetOptimisticBlockStreamResponse,
+    },
+    primitive::v1::RollupId,
+    Protobuf,
+};
+use tonic::{
+    codegen::tokio_stream::{
+        wrappers::ReceiverStream,
+        Stream,
+    },
+    Request,
+    Response,
+    Status,
+};
+use tracing::{
+    debug,
+    info,
+};
+
+use crate::app::OptimisticBlockChannels;
+
+pub(crate) struct OptimisticBlockServer {
+    optimistic_block_channels: OptimisticBlockChannels,
+}
+
+impl OptimisticBlockServer {
+    pub(crate) fn new(optimistic_block_channels: OptimisticBlockChannels) -> Self {
+        Self {
+            optimistic_block_channels,
+        }
+    }
+}
+
+#[async_trait::async_trait]
+impl OptimisticBlockService for OptimisticBlockServer {
+    type GetBlockCommitmentStreamStream =
+        Pin<Box<dyn Stream<Item = Result<GetBlockCommitmentStreamResponse, Status>> + Send>>;
+    type GetOptimisticBlockStreamStream =
+        Pin<Box<dyn Stream<Item = Result<GetOptimisticBlockStreamResponse, Status>> + Send>>;
+
+    async fn get_optimistic_block_stream(
+        self: Arc<Self>,
+        request: Request<GetOptimisticBlockStreamRequest>,
+    ) -> Result<Response<Self::GetOptimisticBlockStreamStream>, Status> {
+        let get_optimistic_block_stream_request = request.into_inner();
+
+        let rollup_id = {
+            let rollup_id = get_optimistic_block_stream_request
+                .rollup_id
+                .ok_or_else(|| Status::invalid_argument("rollup id is required"))?;
+
+            RollupId::try_from_raw(rollup_id)
+                .map_err(|e| Status::invalid_argument(format!("invalid rollup id: {e}")))?
+        };
+
+        let (tx, rx) =
+            tokio::sync::mpsc::channel::<Result<GetOptimisticBlockStreamResponse, Status>>(128);
+
+        let mut optimistic_block_receiver = self
+            .optimistic_block_channels
+            .optimistic_block_sender()
+            .subscribe();
+
+        tokio::spawn(async move {
+            loop {
+                while let Ok(()) = optimistic_block_receiver.changed().await {
+                    let optimistic_block = optimistic_block_receiver
+                        .borrow_and_update()
+                        .clone()
+                        .expect("received an invalid optimistic block");
+
+                    let filtered_optimistic_block =
+                        optimistic_block.to_filtered_block(vec![rollup_id]);
+                    let raw_filtered_optimistic_block = filtered_optimistic_block.into_raw();
+
+                    let get_optimistic_block_stream_response = GetOptimisticBlockStreamResponse {
+                        block: Some(raw_filtered_optimistic_block),
+                    };
+
+                    match tx.send(Ok(get_optimistic_block_stream_response)).await {
+                        Ok(()) => {
+                            debug!("sent optimistic block");
```

Comment:
should we log this? this will log multiple times with each new listener of the stream.